openfl/experimental/workflow/runtime/local_runtime.py

```diff
@@ -593,49 +579,42 @@ def restore_instance_snapshot(self, ctx: Type[FLSpec], instance_snapshot: List[T
                 if not hasattr(ctx, name):
                     setattr(ctx, name, attr)
 
-    def execute_agg_steps(self, ctx: Any, f_name: str, clones: Optional[Any] = None):
-        """
-        Execute aggregator steps until at transition point.
-
-        Args:
-            ctx (Any): The context in which the function is executed.
-            f_name (str): The name of the function to be executed.
-            clones (Optional[Any], optional): Clones if any. Defaults to None.
-        """
-        if clones is not None:
-            f = getattr(ctx, f_name)
-            f(clones)
-        else:
-            not_at_transition_point = True
-            while not_at_transition_point:
-                f = getattr(ctx, f_name)
-                f()
-
-                f, parent_func = ctx.execute_task_args[:2]
-                if aggregator_to_collaborator(f, parent_func) or f.__name__ == "end":
-                    not_at_transition_point = False
-
-                f_name = f.__name__
-
-    def execute_collab_steps(self, ctx: Any, f_name: str):
-        """Execute collaborator steps until at transition point.
+    def _execute_flow(self, flspec_obj: Type[FLSpec]):
+        """Executes and updates the flow with  the final attributes.
 
         Args:
-            ctx (Any): The context in which the function is executed.
-            f_name (str): The name of the function to be executed.
+            flspec_obj: Reference to the FLSpec (flow) object.
         """
-        not_at_transition_point = True
-        while not_at_transition_point:
-            f = getattr(ctx, f_name)
-            f()
-
-            f, parent_func = ctx.execute_task_args[:2]
-            if ctx._is_at_transition_point(f, parent_func):
-                not_at_transition_point = False
+        try:
+            # Execute all Participant (Aggregator & Collaborator) tasks and
+            # retrieve the final attributes
+            # start step is the first task & invoked on aggregator through
+            # runtime.execute_task
+            final_attributes = self._execute_task(
+                flspec_obj,
+                flspec_obj.start,
+            )
+        except Exception as e:
+            if "cannot pickle" in str(e) or "Failed to unpickle" in str(e):
+                msg = (
+                    "\nA serialization error was encountered that could not"
+                    "\nbe handled by the ray backend."
+                    "\nTry rerunning the flow without ray as follows:\n"
+                    "\nLocalRuntime(...,backend='single_process')\n"
+                    "\n or for more information about the original error,"
+                    "\nPlease see the official Ray documentation"
+                    "\nhttps://docs.ray.io/en/releases-2.2.0/ray-core/\
+                    objects/serialization.html"
+                )
+                raise SerializationError(str(e) + msg)
```

Comment:
I suggest to extract this in a private method, to avoid polluting the main `_execute_flow()` with corner cases.